// @aws-sdk/client-route-53/dist-es/commands/GetCheckerIpRangesCommand.js
import { getSerdePlugin } from "@aws-sdk/middleware-serde";
import { Command as $Command } from "@aws-sdk/smithy-client";
import { GetCheckerIpRangesRequestFilterSensitiveLog, GetCheckerIpRangesResponseFilterSensitiveLog, } from "../models/models_0";
import { deserializeAws_restXmlGetCheckerIpRangesCommand, serializeAws_restXmlGetCheckerIpRangesCommand, } from "../protocols/Aws_restXml";
export class GetCheckerIpRangesCommand extends $Command {
    constructor(input) {
        super();
        this.input = input;
    }
    resolveMiddleware(clientStack, configuration, options) {
        this.middlewareStack.use(getSerdePlugin(configuration, this.serialize, this.deserialize));
        const stack = clientStack.concat(this.middlewareStack);
        const { logger } = configuration;
        const clientName = "Route53Client";
        const commandName = "GetCheckerIpRangesCommand";
        const handlerExecutionContext = {
            logger,
            clientName,
            commandName,
            inputFilterSensitiveLog: GetCheckerIpRangesRequestFilterSensitiveLog,
            outputFilterSensitiveLog: GetCheckerIpRangesResponseFilterSensitiveLog,
        };
        const { requestHandler } = configuration;
        return stack.resolve((request) => requestHandler.handle(request.request, options || {}), handlerExecutionContext);
    }
    serialize(input, context) {
        return serializeAws_restXmlGetCheckerIpRangesCommand(input, context);
    }
    deserialize(output, context) {
        return deserializeAws_restXmlGetCheckerIpRangesCommand(output, context);
    }
}
